### PR TITLE
🎉 create etl step update

### PR DESCRIPTION
**Pull request by pabloarosado**
# Overview
This PR:
* Creates a new CLI tool `etl-step-update` that lets the user update one or more steps, optionally updating its dependencies and usages too. It will not only duplicate the files, but also write the new steps and dependencies to the corresponding dag file. New snapshots (while they are not used by any active step) are stored in a temporary dag file (`temp.yaml`).
* Makes some changes to the etl-wizard, to let `etl-step-update` open a tab in the browser and be able to edit metadata more easily. We will probably revert these changes once `etl-step-update` is properly integrated as part of wizard.
* Adds a helper function to update the content of dag files respecting comments (a bit hacky, I'm sure it can be made more efficiently).

# Reviewers
Since this doesn't affect any public data or metadata, I don't think a thorough review is necessary. But it would be good to give the tool a go, to try to find some bugs and edge cases. Feel free to follow the examples in `etl-step-update --help`.

@lucasrodes  why is the help message of this tool formatted so poorly, compared to, e.g. `etl chart-sync --help`?

TODO:
* I will add unit tests while this PR is reviewed.
* I will start looking into the integration of this tool in the etl-wizard ecosystem (may need some of @lucasrodes' advise).

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -13,6 +13,7 @@ etl-wizard = 'apps.wizard.cli:cli'
 fasttrack = 'apps.fasttrack.cli:cli'
 compare = 'etl.compare:cli'
 backport = 'apps.backport.backport:backport_cli'
+etl-step-update = 'apps.step_update.cli:cli'
```
Comment (lucasrodes) [inline pyproject.toml:16]: We are moving away from using `etl-*` like commands. Instead, we should put these under our single client `etl`.

More on the motivation for this here: https://github.com/owid/etl/pull/2293

File: etl/paths.py [hunk 2]
```diff
@@ -5,6 +5,7 @@
 DAG_DIR = BASE_DIR / "dag"
 DAG_FILE = DAG_DIR / "main.yml"
 DAG_ARCHIVE_FILE = DAG_DIR / "archive" / "main.yml"
+DAG_TEMP_FILE = DAG_DIR / "temp.yml"
```
Comment (lucasrodes) [inline etl/paths.py:8]: I don't see this being used in other files. Maybe you forgot to undo this?

**Review by lucasrodes — CHANGES_REQUESTED** [review]
Hey @pabloarosado, thanks for working on this. It is really nice to see a tool in this space flourishing, looks promising!

In general, I am a bit skeptical if a CLI with multiple options is what we want here. Instead, I think that it should just be part of Wizard, without the need to run another CLI from the terminal (`etl-step-update`). In short, I think we need (i) backend implementation and (ii) a Wizard app.

Therefore, I would probably just preserve the pure backend code from `apps/step_udpate/` but remove the CLI (or at least separate the backend from the CLI).


Some other feedback on this particular implementation:

1. Could you try to add this command to our new CLI `etl`? We are moving away of standalone commands (`etl-*` kinda commands). Adding it there will render markdown in the `--help` message nicely, too. You can add it in `apps/cli/__init__.py`. Just add it either to `GROUPS` or `SUBROUPS`, depending if you want it to be a top-command or a subgroup command. I think `etl update` would be nice.
2. We should not have wizard-app-specific arguments in the `etlwiz`/`etl-wizard` calls. Adding more arguments to `etlwiz` should only be done if these affect all (or most) of Wizard apps. In this case, `--data-from-step` and `--snapshot-version` are for a very specific use case. Instead, as we discussed, I think that the metadata to pre-fill Wizard fields should be loaded from a dropdown (or something similar) WITHIN the Wizard app. This way, we also can add different metadata without the need to re-run `wizard` each time. Let's not add these, and think of a better way to do this from the UI.

**Comment by Marigold** [comment]
Code in general looks good, but I'm getting an error when trying to execute `etl-step-update garden/biodiversity/2024-01-25/cherry_blossom --step-version-new 2024-02-20 --dry-run`
```
[1]    8118 segmentation fault  etl-step-update garden/biodiversity/2024-01-25/cherry_blossom  2024-02-20
/Users/mojmir/.pyenv/versions/3.10.0/lib/python3.10/multiprocessing/resource_tracker.py:224: UserWarning: resource_tracker: There appear to be 2 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
```

I'll wait for you and Lucas to settle down on whether it should be in wizard or not and then I'll take another look. (Could you also merge with master, please?)

File: etl/paths.py (hunk 2 above)
Comment (pabloarosado) [reply etl/paths.py:8]: `DAG_TEMP_FILE` is used by `apps/step_update/cli.py`.

File: apps/wizard/cli.py [hunk 3]
```diff
@@ -82,6 +91,12 @@ def cli(phase: Iterable[WIZARD_PHASES], run_checks: bool, dummy_data: bool, port
         args.append("--run-checks")
     if dummy_data:
         args.append("--dummy-data")
+    if data_from_step:
+        args.append("--data-from-step")
+        args.append(data_from_step)
+    if snapshot_version:
+        args.append("--snapshot-version")
+        args.append(snapshot_version)
```
[the remark this answers is not in the document]
Comment (pabloarosado) [reply apps/wizard/cli.py:99]: I don't like this solution either, but it's a temporary workaround while the tool is not fully integrated in wizard. We could remove it, but it lets us already edit snapshot metadata files more easily. Do you think this workaround does any harm? (Apart from leaving bad code behind if we forget about it).

File: apps/wizard/etl_steps/snapshot.py [hunk 4]
```diff
@@ -58,6 +59,25 @@
...
+        "origin.date_published": metadata.origin.date_published,  # type: ignore
+        "origin.producer": metadata.origin.producer,  # type: ignore
+        "origin.citation_full": metadata.origin.citation_full,  # type: ignore
+        "origin.url_main": metadata.origin.url_main,  # type: ignore
+        "origin.date_accessed": metadata.origin.date_accessed,  # type: ignore
+    }
+    if APP_STATE.args.snapshot_version:
+        dummy_values["snapshot_version"] = APP_STATE.args.snapshot_version
+        dummy_values["origin.date_accessed"] = APP_STATE.args.snapshot_version
+
```
[the remark this answers is not in the document]
Comment (pabloarosado) [reply apps/wizard/etl_steps/snapshot.py:80]: It's a temporary solution. Do you think it's risky for some reason?

File: apps/wizard/cli.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lucasrodes) [reply apps/wizard/cli.py:99]: I think that we can have something in Wizard this week, which would make this implementation not really needed.

so I'm in favor of just adding it there. I am happy to help out there!

File: apps/wizard/etl_steps/snapshot.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lucasrodes) [reply apps/wizard/etl_steps/snapshot.py:80]: not risky at all. 

On top of what I mentioned in my previous comments, I just think that we can just have it on Wizard this week (which makes this not really needed). I am happy to help there!

**Comment by pabloarosado** [comment]
> Code in general looks good, but I'm getting an error when trying to execute `etl-step-update garden/biodiversity/2024-01-25/cherry_blossom --step-version-new 2024-02-20 --dry-run`
> 
> ```
> [1]    8118 segmentation fault  etl-step-update garden/biodiversity/2024-01-25/cherry_blossom  2024-02-20
> /Users/mojmir/.pyenv/versions/3.10.0/lib/python3.10/multiprocessing/resource_tracker.py:224: UserWarning: resource_tracker: There appear to be 2 leaked semaphore objects to clean up at shutdown
>   warnings.warn('resource_tracker: There appear to be %d '
> ```
> 
> I'll wait for you and Lucas to settle down on whether it should be in wizard or not and then I'll take another look. (Could you also merge with master, please?)

Hey @Marigold that's a really strange error. I just ran it and it works for me (with the only caveat that the tool requires the `data://` at the beginning of the step):
Screenshot 2024-02-20 at 14 34 45

**Comment by lucasrodes** [comment]
As you decide, Pablo. It won't make any harm. 

I've just been investing some time cleaning and organizing the tooling and have some strong opinions on it.

**Comment by pabloarosado** [comment]
> As you decide, Pablo. It won't make any harm.
> 
> I've just been investing some time cleaning and organizing the tooling and have some strong opinions on it.

OK, I'll remove the `--edit-metadata` option.

**Comment by lucasrodes** [comment]
LGTM, Pablo! thanks for adding the suggestions & tests. Let's continue working this week.

**Review by lucasrodes — APPROVED** [review]

**Comment by lucasrodes** [comment]
I forgot to submit my comments when I reviewed earlier today ー  they are not important, so we can tweak this later.

**Pull request merged**